### DISCUSSION TITLE
Wix4 - Triggerring custom action on feature enabled

### DISCUSSION THREAD

**Discussion by AntonyNifosi** [wixtoolset/issues#7882] [Questions]
I would like to trigger a custom action when the user will install a feature.
Let's say for example you have your main program and a Visual Redist feature, if you select to install this feature it will execute the vcredist.exe installer.

I did like this :

```XML
    <SetProperty Id="InstallVCRedist" Value="&quot;[INSTALLFOLDER]\vcredist.exe&quot;"
      Sequence="execute"
      Before="InstallVCRedist" />
    <CustomAction Id="InstallVCRedist" BinaryRef="Wix4UtilCA_X64" DllEntry="WixQuietExec64"
      Return="check" Impersonate="no" Execute="deferred" />

    <InstallExecuteSequence>
      <Custom Action="InstallVCRedist" Before="InstallFinalize" Condition="Vcredist=3"/>
    </InstallExecuteSequence>
````
But it doesn't seem to work, if I remove the "Condition=...." it works properly, so it's problably the way I am trying to access to the Vcredist value which is wrong.

Anyone knows the good way to do that ?

Thank's in advance for your answers !

**Comment by barnson** [comment]
https://learn.microsoft.com/en-us/windows/win32/msi/conditional-statement-syntax but that won't work because you can't install one MSI from another so you need https://wixtoolset.org/docs/tools/burn/.

**Reply by AntonyNifosi** [reply]
But like I said if I remove the condition, the custom action does trigger correctly and my vcredist.exe installer is correctly executed and the Visual C++ Redistribuable are correctly installed.

I don't really understand why you say it's not possible, my only problem is the condition, not the triggering of the custom action.

**Reply by wmanning** [reply]
Maybe the redistribuable is not a MSI package, not sure.  A MSI package will prevent other MSIs from being run while it is being installed.

**Comment by chrpai** [comment]
@AntonyNifosi MSI has a mutex that limits 1 installation execute sequence at a time or 1 UI sequence per process.  It is technically possible for an MSI to shell out to another MSI from it's UI sequence but this isn't a best practice because 1) System state changes shouldn't be happening prior to InstallInitialize  and 2)  this precudes your installer from running silently.  However I note that your wix code shows it being called from the execute sequence.  That is 100% not possible.   VCRedist should return 1618 and fail.

https://learn.microsoft.com/en-us/windows/win32/msi/-msiexecute-mutex

**Reply by AntonyNifosi** [reply]
I managed to do it by getting the return value of vcredist and checking if it's 1618, if it is I just exit 0 ... But I agree this is not a really "good" way of doing it; it works like I want but doesn't seem the best approach.
So how I should do it properly ?
My goal is to propose the user to install Visual Redistribuables (with feature I suppose) when he is installing my product.

**Reply by AntonyNifosi** [reply]
Can I use the ExePackage element to install vcredist, is it better than my InstallSequence ?

**Comment by chrpai** [comment]
Every VCRedist package I've ever seen has included MSI inside.  I just downloaded the latest and it's still a wix bundle with MSIs inside.

**Comment by chrpai** [comment]
1618 means VCRedist wasn't installed.  That's not a 0 ERROR_SUCCESS.  The correct way to do it is use a bundle like Bob suggested. Yes, ExePackage inside of a bundle is better.
